File: FactorAnalysis/EMAlg.m
function [hlambda, hpsi, ind_llh] = EMAlg(X,k)

p = size(X,1);
n = size(X,2);

Cxx = (1/n)*(X*X'); 

% Initial guess of lambda and psi;
[hlambda, hpsi] = factoran(X',k, 'maxit', 1000);

% Tolerance parameters for EM alg
tol = 1e-15;
maxiter = 500;
llh = -inf(1,maxiter);

for iter = 2:maxiter
    
    % log-likelihood
    V = chol(diag(hpsi)+hlambda*hlambda');  % diag(hpsi) + hlambda*hlambda' = V'V;
    logdetV = log(det(V));                  % the det(V) is the square of the det of sigmax;
    loglike = -(1/2)*(p*log(2*pi)+2*logdetV+diag((X'/V)/(V')*X));
     
    llh(iter) = -(n/2)*(p*log(2*pi)+2*logdetV+(1/n)*trace((X'/V)/(V')*X));
    if abs(llh(iter) - llh(iter-1)) < tol*abs(llh(iter-1)); break; end
        
    % EM algorithm
    hbeta   = (hlambda'/V)/V';
    chol_Ezz = chol(eye(k) - hbeta*hlambda + hbeta*Cxx*hbeta');
    
    hlambda = ((Cxx*hbeta')/chol_Ezz)/chol_Ezz';
    hpsi    = diag((eye(p) - hlambda*hbeta)*Cxx);
    
end

ind_llh = loglike; %log likelihood for each individual

end

File: FactorAnalysis/R_code/EMAlg.R
EMAlg <- function(X, k){
  
  # data dimensions
  n = nrow(X)
  p = ncol(X)
  
  # sample mean and covariance matrix
  nu = apply(X, 2, mean)
  X_centered = sweep(X, 2, nu, "-")
  Cxx = 1/n * t(X_centered) %*% X_centered
  
  # initial guess of lambda and psi
  intl_guess <- factanal(X, k)
  lambda <- intl_guess$loadings
  psi <- intl_guess$uniquenesses
  
  # tolerance parameters
  tol = 1e-6
  maxiter = 1e3
  
  # EM algorithm
  obj = rep(-Inf, maxiter)
  for (iter in 2:maxiter){
    
    # calculate log-likelihoods
    sigma = lambda %*% t(lambda) + diag(psi)
    V = chol(sigma)
    inv_V = solve(V)
    logdetV = log(det(V))
    Z = sweep(X, 2, nu, "-") %*% inv_V
    ind_lik = -(1/2)*(p*log(2*pi) + 2*logdetV + apply(Z, 1, function(y) t(y) %*% y))
    
    # calculate value for objective function
    obj[iter] = sum(ind_lik)
    
    # check for convergence
    if(abs(obj[iter] - obj[iter-1]) < tol*abs(obj[iter-1])){break}
    
    # estimate beta
    beta = (t(lambda) %*% inv_V) %*% t(inv_V)
    
    # estimate lambda and psi
    Ezz = diag(k) - beta %*% lambda + beta %*% Cxx %*% t(beta)
    chol_Ezz = chol(Ezz)
    inv_chol_Ezz = solve(chol_Ezz)
    lambda = ((Cxx %*% t(beta)) %*% inv_chol_Ezz) %*% t(inv_chol_Ezz)
    psi = diag((diag(p) - lambda %*% beta) %*% Cxx)
   
    if(iter==maxiter){message(
      paste('maxiter reached; increase maxiter', abs(obj[iter] - obj[iter-1])/abs(obj[iter-1])))}
    
  }
  
  # rescale estimates
  total_var = apply(lambda^2, 1, sum) + psi
  lambda = apply(lambda, 2, function(x) x / sqrt(total_var))
  psi = psi / total_var
  nu =  nu / sqrt(total_var)
  
  # rotate loadings
  lambda = varimax(lambda)
  
  output = list(nu = nu, lambda = lambda$loadings, psi = psi, ind_lik = ind_lik)
  return(output)
}